File: algorithms/java/src/test/java/leetcode/LC214ShortestPalindrome.java
package leetcode.lc214_shortest_palindrome;

import org.junit.jupiter.api.Test;


import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/shortest-palindrome/
 * #hard
 */
public final class LC214ShortestPalindrome {
    public final class Solution {
        public String shortestPalindrome(final String s) {
            if (s == null || s.length() == 0) {
                return s;
            }
            String A = s + new StringBuffer(s).reverse().toString();
            int[] cont = new int[A.length()];
            cont[0] = 0;
            for (int i = 1; i < A.length(); i++) {
                int index = cont[i - 1];
                while (index > 0 && A.charAt(index) != A.charAt(i)) {
                    index = cont[index - 1];
                }
                cont[i] = index + (A.charAt(index) == A.charAt(i) ? 1 : 0);
            }
            return new StringBuilder(s.substring(cont[cont.length - 1], s.length())).reverse().toString() + s;
        }
    }

    @Test
    public void test_aacecaaa() throws Exception {
        assertEquals("aaacecaaa", new Solution().shortestPalindrome("aacecaaa"));
    }

    @Test
    public void test_abcd() throws Exception {
        assertEquals("dcbabcd", new Solution().shortestPalindrome("abcd"));
    }
}

File: algorithms/code/leetcode/lc214_shortest_palindrome/lc214_shortest_palindrome.py
#!/usr/bin/env python3
# https://leetcode.com/problems/shortest-palindrome/

import unittest


class Solution:
    def shortestPalindrome(self, s: str) -> str:
        if s is None or len(s) == 0:
            return s
        A = s + s[::-1]
        cont = [0] * len(A)
        for i in range(1, len(A)):
            index = cont[i - 1]
            while index > 0 and A[index] != A[i]:
                index = cont[index - 1]
            cont[i] = index + (1 if A[index] == A[i] else 0)
        return s[cont[- 1]:][::-1] + s


class TestCode(unittest.TestCase):
    def test_aacecaaa(self):
        self.assertEqual(
            'aaacecaaa', Solution().shortestPalindrome('aacecaaa'))

    def test_abcd(self):
        self.assertEqual('dcbabcd', Solution().shortestPalindrome('abcd'))

    def test_nothing(self):
        self.assertIsNone(Solution().shortestPalindrome(None))
        self.assertEqual('', Solution().shortestPalindrome(''))
